Srour

Ali-Alzyoud opened the pull request.

AbedSrour committed:
Add undo/redo example
Update README.md
Add .gitignore

AbedSrour committed:
Update README.md

AbedSrour committed:
Undo/Redo use eina_list for history

Ali-Alzyoud committed:
add comments

Ali-Alzyoud commented on Undo_Redo_Example.c:
```diff
@@ -0,0 +1,231 @@
+static void
+_ui_text_undo_clicked_cb(void *data, const Efl_Event *event EINA_UNUSED)
+{
+   Eo *ui_text = data;
```
How to force Event programmatically  (addition to Ctrl+Z fire event on button click)

tasn replied:
You just trigger the callback yourself. So trigger the "undo,requested". Same with redo.

Ali-Alzyoud replied:
Sorry, But I do not know how can user trigger event by him self ? (Is this feature available in EFL)

Ali-Alzyoud commented on Undo_Redo_Example.c:
```diff
@@ -0,0 +1,231 @@
+   // Register callback when text changed to track new changes
+   efl_event_callback_add(ui_text, EFL2_CANVAS_TEXT_EVENT_CHANGED, _ui_text_changed_cb, ui_text);
+
+   // Register callback when redu event called to recall future state (if exists)
```
(1) EFL2_TEXT_RAW_EDITABLE_EVENT_REDO_REQUEST will be available for efl.ui.text ?
(2) I think user expect event name be like  EFL2_UI_TEXT_EVENT_REDO_REQUEST ?

tasn replied:
Yes it will be.

Fair point about the naming though, I added a comment and we'll have to fix the naming.

Ali-Alzyoud replied:
Other Question please 
EFL2_TEXT_RAW_EDITABLE_CURSOR_CHANGED_MANUAL,  what is **MAUNAL** means ? if user change main cursor position programmatically, will ui_text fire this event ?

Also

Which one is better listen to ui_text.main_cursor **onchange** or listen to ui_text **CURSOR_CHANGED_MANUAL** ? or they are the same

tasn commented:
As I said on the other thread:

As for the change: it doesn't look like you're using the change_info structure, which is a key element for undo/redo.

AbedSrour committed:
Update undo redo Example
Use Change Info instead of whole text
New changes are based on Ecrire text editor

Ali-Alzyoud commented:
Example updated

tasn commented on Undo_Redo_Example.c:
```diff
@@ -0,0 +1,298 @@
+   if ((inf->insert && undo) || (!inf->insert && !undo))
+     {
+        Efl_Text_Cursor  *end;
+        end = efl2_ui_text_cursor_new(ui_text);
```
There is no such function.

tasn replied:
Also, maybe consider caching this cursor? No reason to allocate it every time.
You can use e.g. efl_named_data_set or whatever it's name is to just cache it on the object and use the key "_undo/redo_cursor" or whatever you want. Just an example.

Ali-Alzyoud replied:
>  There is no such function.

Do you mean efl2_ui_text_cursor_new? Please check efl.ui.text.eo

tasn replied:
Have it open in front of me... Maybe you're based on an old version?

The correct way for creating new cursors is a bit weird at the moment, look at the tests (recent changes in the git log) to see how it's done.

Ali-Alzyoud replied:
```
cursor_new {
         [[Creates and returns a new cursor for the text.]]
         return: ptr(Efl.Text_Cursor_Cursor); [[Text cursor]]
      }
```
I see this inside https://github.com/Enlightenment/efl/blob/devs/tasn/ifaces/src/lib/elementary/efl2_ui_text.eo

tasn replied:
I was indeed looking at the wrong one! I was looking at canvas_text. Anyhow, this one is an old leftover as you can see from the fact that it doesn't use the correct return value (Efl.Text_Cursor_Cursor is the old type). I'll push a fix in a bit.

tasn reviewed (changes requested):
Great job! Overall it looks great, I just had a few minor comments based on your questions.

Btw, have you looked at how it's done in ecrire? Maybe you can draw some inspiration from there: https://git.enlightenment.org/apps/ecrire.git/tree/src/bin/main.c

Ali-Alzyoud commented:
> Btw, have you looked at how it's done in ecrire? Maybe you can draw some inspiration from there: https://git.enlightenment.org/apps/ecrire.git/tree/src/bin/main.c

Yes we did
How about adding this feature inside our text object? I think leaving implementation to user is not very helpful for him.

tasn commented:
> > Btw, have you looked at how it's done in ecrire? Maybe you can draw some inspiration from there: https://git.enlightenment.org/apps/ecrire.git/tree/src/bin/main.c
> 
> Yes we did
> How about adding this feature inside our text object? I think leaving implementation to user is not very helpful for him.

You already brought this up. I 100% agree, this should (and will be) inside Efl.Ui.Text.
